streaming/src/data_writer.cc

```diff
@@ -447,5 +431,54 @@ void DataWriter::EmptyMessageTimerCallback() {
   }
 }
 
+void DataWriter::NotifyConsumedByRefresh(ProducerChannelInfo &channel_info) {
+  // Refresh current downstream consumed seq id.
+  channel_map_[channel_info.channel_id]->RefreshChannelInfo();
+  // Notify the consumed information to local channel.
+  NotifyConsumedItem(channel_info, channel_info.queue_info.consumed_seq_id);
+}
+
+void DataWriter::NotifyConsumedItem(ProducerChannelInfo &channel_info, uint32_t offset) {
+  if (offset > channel_info.current_seq_id) {
+    STREAMING_LOG(WARNING) << "Can not notify consumed this offset " << offset
+                           << " that's out of range, max seq id "
+                           << channel_info.current_seq_id;
+  } else {
+    channel_map_[channel_info.channel_id]->NotifyChannelConsumed(offset);
+  }
+}
+
+void DataWriter::FlowControlTimer() {
+  std::chrono::milliseconds MockTimer(
+      runtime_context_->GetConfig().GetEventDrivenFlowControlInterval());
+  while (true) {
+    if (runtime_context_->GetRuntimeStatus() != RuntimeStatus::Running) {
+      return;
+    }
+    for (const auto &output_queue : output_queue_ids_) {
+      if (runtime_context_->GetRuntimeStatus() != RuntimeStatus::Running) {
+        return;
+      }
+      ProducerChannelInfo &channel_info = channel_info_map_[output_queue];
+      if (!channel_info.flow_control) {
+        continue;
+      }
+      if (!flow_controller_->ShouldFlowControl(channel_info)) {
+        channel_info.flow_control = false;
+        Event event{&channel_info, EventType::FlowEvent,
+                    channel_info.writer_ring_buffer->IsFull()};
+        event_service_->Push(event);
+        ++channel_info.flow_control_cnt;
+      }
+    }
+    std::this_thread::sleep_for(MockTimer);
+  }
+}
+
+void DataWriter::GetOffsetInfo(
```

Comment:
Is this function just for test?

Reply:
> Is this function just for test?

For checkpoint, but it's can be used in testing.